keps/0019-package-api-versioning.md

```diff
@@ -0,0 +1,72 @@
+---
+kep-number: 19
+title: Versioning of KUDO API in Operator Packages
+authors:
+  - "@nfnt"
+owners:
+  - "@nfnt"
+creation-date: 2019-11-05
+last-updated: 2019-11-05
+status: provisional
+---
+
+# Versioning of KUDO API in Operator Packages
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+By adding a KUDO API version to package indexes, we will avoid that changes to the operator schema will affect existing KUDO versions and package definitions.
+
+## Motivation
+
+Operator packages are developed by defining the operator using YAML files. The schema to use for these files is defined by KUDO. While the KUDO team tries to keep the schema definitions consistent, there can be situations where the schema has to be changed. We need to ensure that packages are in sync with these changes while at the same time allowing existing packages to work with existing versions of KUDO.
+
+### Goals
+
+* Allow users to install packages from a package repository that provides packages for multiple KUDO API versions
+* Reject packages that don't have a matching KUDO API version
+
+### Non-Goals
+
+* Handle operator dependencies on specific Kubernetes versions
+
+## Proposal
+
+Through the `apiVersion` field in an `operator.yaml` it is already guaranteed that the operator will match a specific API version of KUDO. However, this field isn't taken into account when installing packages from a repository. KUDO will try to install the latest version of the package even if the `apiVersion` doesn't match. This can be mitigated by adding the packages `apiVersion` to the repository index and filtering out all packages that don't have a matching version. This way, older versions of packages can still be part of the repository and keep working with older versions of KUDO.
```

Comment:
just couple of questions:
- what does it mean for the version to match? will we be releasing a matrix with every version that will say what api versions are supported and then we'll have the same matrix for evaluation in the kudo code?
- will all the api versions be in sync? I actually don't think so but we'll probably need to discuss that. I can see params going from v1beta1 to v1beta2 because we need to change something there but we won't need to change anything in operator so that stays. or do all of them now have to move in the same pace? both development api versions as well as server-side? I guess we need to clear this out and probably capture in this KEP as well

Reply:
Currently, KUDO (controller as well as CLI) already only supports a specific `apiVersion`, the one it has been built with. This is currently `v1beta1`. It isn't in the scope of the KEP to support multiple API versions in the CLI. Hence, API versions need to stay in sync between controller and CLI.

Reply:
oh ok, I guess I understood the kep as a general versioning kep where we discuss a versioning strategy overall, since it's all connected and we don't have any so implementing that for packages without the strategy seems like solving just an implementation detail. Maybe we need to start one more KEP then :)

Reply:
Let me improve the title and description of this KEP to narrow down the scope a bit. The main purpose is to prevent that deployed versions of KUDO suddenly aren't able to install any operators after we release a new version of KUDO with an updated API.

Reply:
I changed the scope of the KEP to include these questions. Resolving these conversations to focus on the updated KEP.